sklearn/model_selection/_split.py

```diff
@@ -729,8 +733,12 @@ def split(self, X, y=None, groups=None):
         test_starts = range(test_size + n_samples % n_folds,
                             n_samples, test_size)
         for test_start in test_starts:
-            yield (indices[:test_start],
-                   indices[test_start:test_start + test_size])
+            if self.max_train_size > 0 and self.max_train_size < test_start:
+                yield (indices[test_start - self.max_train_size:test_start],
```

Comment:
The check ensures that only when the current fold's index exceeds the `max_train_size` will this be yielded. Else, it'll revert back to the default. I have updated the tests to capture what I felt you were mentioning.